### PR TITLE
DAOS-623 ci: Setting test tags and skipping enhancements

### PR DESCRIPTION
**Pull request by brianjmurrell**
Now `Test-tag:` is the tag(s) to be applied to all Functional test
stages.
- This replaces the default of `pr` when used and only runs the tests
  defined by the tag(s).
- Add `pr` to this tag to also run the regular `pr` tests or better yet,
  use `Features:` instead of `Test-tag:` to run tests in addition to the
  regular `pr` tests.

Stages that don't have any tests matching the selected tags are now
automatically skipped.

Add a `Quick-Functional:` commit pragma to do only functional testing as
quickly as possible with the minimum of builds required.

Signed-off-by: Brian J. Murrell <brian.murrell@intel.com>

### CODE REVIEW

**Review by JohnMalmberg — APPROVED** [review]

File: Jenkinsfile [hunk 1]
```diff
@@ -329,49 +334,84 @@ boolean skip_build() {
...
+        for (String tag in tags.split(" ")) {
+            newtags.add(tag + ",-hw")
+        }
+        tags += ",-hw"
+    } else {
+        if (size == "medium") {
+            size += ",ib2"
+        }
+        for (String tag in tags.split(" ")) {
+            newtags.add(tag + ",hw," + size)
```
[the remark this answers is not in the document]
Comment (brianjmurrell) [reply Jenkinsfile:386]: Fixed.

**Comment by brianjmurrell** [comment]
This is ready to be reviewed if anyone could oblige please and thanks.

The only failure is a newly filed segfault in the IO server: [DAOS-6597](https://jira.hpdd.intel.com/browse/DAOS-6597)

File: src/tests/ftest/launch.py [hunk 2]
```diff
@@ -154,70 +160,81 @@ def set_test_environment(args):
...
+                            if ioerror.errno == errno.EINVAL:
+                                speed = 1000
+                            else:
+                                raise
+                    print(
+                        "  - {0:<5} (speed: {1:>6} state: {2})".format(
+                            device, speed, state))
+                    # Only include the first active interface for each speed
+                    # First # is determined by an alphabetic sort: ib0 will be
+                    # checked before ib1
```
Comment (phender) [inline src/tests/ftest/launch.py:199]: Small correction on the updated indent:
```suggestion
                    # Only include the first active interface for each speed -
                    # first is determined by an alphabetic sort: ib0 will be
                    # checked before ib1
```

File: src/tests/ftest/launch.py [hunk 3]
```diff
@@ -106,8 +106,14 @@ def get_build_environment():
...
-    with open(build_vars_file) as vars_file:
-        return json.load(vars_file)
+    try:
+        with open(build_vars_file) as vars_file:
+            return json.load(vars_file)
+    except ValueError:
+        return json.loads('{{"PREFIX": "{}"}}'.format(os.getcwd()))
+    except IOError as error:
+        if error.errno == errno.ENOENT:
+            return json.loads('{{"PREFIX": "{}"}}'.format(os.getcwd()))
```
Comment (phender) [inline src/tests/ftest/launch.py:116]: Should these alternatives only be acceptable in certain circumstances?  In other words is there a situation where if `../../.build_vars.json` does not exist the correct action would be to report an error?

If not would it make sense to bypass all of this work and just use `os.getcwd()` to define the prefix?  Possibly something to address later on.

File: src/tests/ftest/get_remote_files.sh [hunk 4]
```diff
@@ -269,6 +269,11 @@ if ! ret=$(check_files_input "${FILES_TO_PROCESS}"); then
         exit 1
     elif [[ ${ret} -eq 2 ]]; then
         echo "No files matched ${FILES_TO_PROCESS}. Nothing to do."
+        if [ -n "${ARCHIVE_DEST}" ]; then
```
Comment (phender) [inline src/tests/ftest/get_remote_files.sh:272]: This should also take into account whether or not `$VERBOSE` is set as there won't be a log file to copy if it is set.  Double check the bash syntax, but something like this:
```suggestion
        if ! "${VERBOSE}" && [ -n "${ARCHIVE_DEST}" ]; then
```

**Review by phender — CHANGES_REQUESTED** [review]

File: src/tests/ftest/launch.py (hunk 3 above)
Comment (brianjmurrell) [reply src/tests/ftest/launch.py:116]: > Should these alternatives only be acceptable in certain circumstances?

Such as `args.list == true`?

> If not would it make sense to bypass all of this work and just use `os.getcwd()` to define the prefix?

I figured if the file does happen to exist, might as well use it.

File: src/tests/ftest/get_remote_files.sh (hunk 4 above)
Comment (brianjmurrell) [reply src/tests/ftest/get_remote_files.sh:272]: Done

File: src/tests/ftest/launch.py (hunk 3 above)
Comment (brianjmurrell) [reply src/tests/ftest/launch.py:116]: Done

**Review by phender — DISMISSED** [review]

**Comment by brianjmurrell** [comment]
Second request for reviews if anyone could please oblige.

@phender Can I get a re-review from you?  I had a copyright merge conflict I had to resolved in bd5fe1a.

**Review by phender — APPROVED** [review]

**Review by JohnMalmberg — APPROVED** [review]

**Comment by brianjmurrell** [comment]
@daos-stack/daos-gatekeeper: just waiting on hardware runs for the most recent patch, but that patch was just a master merge to resolve a copyright conflict.  the previous run ran all stages and only had the known `daos_test -m` failure.

File: src/tests/ftest/launch.py [hunk 5]
```diff
@@ -89,11 +89,21 @@ def get_build_environment():
...
         "../../.build_vars.json")
-    with open(build_vars_file) as vars_file:
-        return json.load(vars_file)
+    try:
+        with open(build_vars_file) as vars_file:
+            return json.load(vars_file)
+    except ValueError:
+        if not args.list:
+            raise
+        return json.loads('{{"PREFIX": "{}"}}'.format(os.getcwd()))
```
[the remark this answers is not in the document]
Comment (brianjmurrell) [reply src/tests/ftest/launch.py:98]: It's satisfying the caller's need to return valid JSON with the `PREFIX` key present.  The file this usually comes from is not present until a build is complete, but this is a case of wanting `launch.py` to operate (for limited use cases) in the source tree without doing a build.

**Pull request merged**